### PR TITLE
Update documentation for BuildHasherDefault

### PR DESCRIPTION
**Pull request by adrianbrink**
This is my proposed fix to this.

This is my first PR to such a big project, so any feedback is welcome. 

Also, could you point me to where more help is needed with the docs?

r? @steveklabnik 
#31242

### CODE REVIEW

File: src/libcore/hash/mod.rs [hunk 1]
```diff
@@ -271,21 +273,21 @@ impl<H: Default + Hasher> BuildHasher for BuildHasherDefault<H> {
...
-impl<H> Default for BuildHasherDefault<H> {
-    fn default() -> BuildHasherDefault<H> {
+impl<H> Clone for BuildHasherDefault<H> {
+    fn clone(&self) -> BuildHasherDefault<H> {
         BuildHasherDefault(marker::PhantomData)
     }
 }
 
-//////////////////////////////////////////////////////////////////////////////
+/// ///////////////////////////////////////////////////////////////////////////
```
Comment (GuillaumeGomez) [inline src/libcore/hash/mod.rs]: Why this change?

File: src/libcore/hash/mod.rs [hunk 2]
```diff
@@ -366,7 +368,7 @@ mod impls {
         );
     }
 
-    impl_hash_tuple! {}
+    impl_hash_tuple!{}
```
Comment (GuillaumeGomez) [inline src/libcore/hash/mod.rs]: Why this change?

File: src/libcore/hash/mod.rs (hunk 1 above)
Comment (adrianbrink) [inline src/libcore/hash/mod.rs]: I rearranged the functions a bit to make more sense to me, but I've updated and changed it back.

File: src/libcore/hash/mod.rs (hunk 2 above)
Comment (adrianbrink) [inline src/libcore/hash/mod.rs]: I think rustfmt for Atom caused this and I have fixed it now.

**Comment by adrianbrink** [comment]
Damn it. The last commit wasn't supposed to go here. I wanted to open a new PR for that. Could someone link me a resource on how to achieve that?

File: src/libcore/ops.rs [hunk 3]
```diff
@@ -1822,6 +1822,15 @@ impl<Idx: PartialOrd<Idx>> RangeToInclusive<Idx> {
 ///     assert_eq!('a', *x);
 /// }
 /// ```
+
+/// Generally the `Deref` trait is not used to access a field within a single field
+/// struct. More commonly it is used to with box types such as `Rc<T>` or `Box<T>`. This allows
+/// the box types to behave like references to T.
+///
+/// ```
+/// An example here would be nice.
```
Comment (GuillaumeGomez) [inline src/libcore/ops.rs]: You can add it yourself, I can't let this get merged. :stuck_out_tongue:

File: src/libcore/ops.rs [hunk 4]
```diff
@@ -1822,6 +1822,15 @@ impl<Idx: PartialOrd<Idx>> RangeToInclusive<Idx> {
...
+/// ````
+
```
Comment (GuillaumeGomez) [inline src/libcore/ops.rs]: Extra empty line.

File: .gitignore [hunk 5]
```diff
@@ -97,3 +97,5 @@ tmp.*.rs
 version.md
 version.ml
 version.texi
+src/libcore/Cargo.lock
```
Comment (GuillaumeGomez) [inline .gitignore]: I'm not sure this change is okay. I don't know what rust core people think about this.

cc @rust-lang/core

Comment (steveklabnik) [inline .gitignore]: Yes, this file should not be changed at all.
